File: C++/155. Min Stack.cpp
#include <bits/stdc++.h>

using namespace std;

// normal way
class MinStack
{
    list<int> stack;
    int min_val;

public:
    MinStack() : min_val(INT_MAX) {}

    void push(int val)
    {
        min_val = min(min_val, val);
        stack.push_back(val);
    }

    void pop()
    {
        int val = stack.back();
        stack.pop_back();
        if (val == min_val)
        {
            min_val = INT_MAX;
            for (int x : stack)
                min_val = min(x, min_val);
        }
    }

    int top()
    {
        return stack.back();
    }

    int getMin()
    {
        return min_val;
    }
};

// good way
class MinStack
{
    list<pair<int, int>> stack;

public:
    MinStack() {}

    void push(int val)
    {
        int min_val = min(this->getMin(), val);
        stack.push_back({val, min_val});
    }

    void pop()
    {
        pair val = stack.back();
        stack.pop_back();
    }

    int top()
    {
        return stack.back().first;
    }

    int getMin()
    {
        if (stack.empty())
            return INT_MAX;
        return stack.back().second;
    }
};

File: Java/155. Min Stack.java

import java.util.ArrayList;
import java.util.List;

class MinStack1 {

    int min;
    List<Integer> stack;

    public MinStack1() {
        min = Integer.MAX_VALUE;
        stack = new ArrayList<>();
    }

    public void push(int val) {
        min = Math.min(min, val);
        stack.add(val);
    }

    public void pop() {
        int val = stack.getLast();
        stack.removeLast();
        if (val == min) {
            min = Integer.MAX_VALUE;
            for (int ele : stack)
                min = Math.min(min, ele);
        }
    }

    public int top() {
        return stack.getLast();
    }

    public int getMin() {
        return min;
    }
}

class MinStack {

    List<Integer> stack;
    List<Integer> minStack;

    public MinStack() {
        stack = new ArrayList<>();
        minStack = new ArrayList<>();
    }

    public void push(int val) {
        stack.add(val);
        minStack.add(Math.min(val, getMin()));
    }

    public void pop() {
        stack.removeLast();
        minStack.removeLast();
    }

    public int top() {
        return stack.getLast();
    }

    public int getMin() {
        if (minStack.isEmpty())
            return Integer.MAX_VALUE;
        return minStack.getLast();
    }
}